src/go/k8s/pkg/resources/statefulset.go

```diff
@@ -80,19 +81,55 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
 
 	r.LastObservedState = &sts
 
-	// Ensure StatefulSet #replicas equals cluster requirement.
-	if sts.Spec.Replicas != nil && r.pandaCluster.Spec.Replicas != nil && *sts.Spec.Replicas != *r.pandaCluster.Spec.Replicas {
-		r.logger.Info(fmt.Sprintf("StatefulSet %s has replicas set to %d but need %d. Going to update", r.Key().Name, *sts.Spec.Replicas, *r.pandaCluster.Spec.Replicas))
-
-		sts.Spec.Replicas = r.pandaCluster.Spec.Replicas
+	updated := Update(&sts, r.pandaCluster, r.logger)
+	if updated {
 		if err := r.Update(ctx, &sts); err != nil {
-			return fmt.Errorf("failed to update StatefulSet: %w", err)
+			return fmt.Errorf("failed to Update StatefulSet: %w", err)
 		}
 	}
 
 	return nil
 }
 
+// Update ensures StatefulSet #replicas and resources equals cluster requirements.
+func Update(
```

Comment:
1. Why this `Update` function doesn't have pointer receiver to `StatefulSetResource`?
2. Why this function isn't added to the `Resource` interface?

Isn't this premature optimisation? Shouldn't we relay on Ensure public API?

Reply:
I guess this is kind of OOP vs FP discussion. If I was about to stick with this being in the OOP style (although there's no true OOP in golang) to be able to test this behavior I would have to mock the client while with this I am able to get a function that accepts all its dependencies as func args and then I can easily test its outputs.

If there's strong opposition against this I can try to refactor.

Reply:
I would like to stick to our public API defined in Resource interface (the `Ensure` function). The tests could be reorganised to ease process of creating the mocks.